src/potdata/sampler.py

```diff
@@ -175,10 +182,71 @@ def __init__(
         self.reachable_ratio = reachable_ratio
         self.core_ratio = core_ratio
 
-        self._indices: list[int] = None
+        self._indices: list[int] = []
 
         np.random.seed(seed)
 
+    def _get_soap_vectors(self, data: list[Structure]) -> list[np.ndarray]:
+        """Convert structures to SOAP vectors.
+           Take Li atom in Li3YCl6 as an example."""
+        
+        # Set up the SOAP descriptor
+        soap_desc = SOAP(
+            species=["Li", "Y", "Cl"],
+            periodic=True, # Set periodicity to True for a periodic system
+            r_cut=5.0,
+            n_max=8,
+            l_max=6,
+            sigma=0.1
+        )
+
+        # Read the trajectory file
+        traj = Trajectory(traj_file)
+
+        Li_indices = [i for i, a in enumerate(traj[0]) if a.symbol == 'Li']
+
+        # Extract the relevant steps, select every 5th in the last 6000 steps
+        relevant_steps = traj[-6000::5]
+
+        # Compute the SOAP vectors for the selected Li atoms at each step
+        soap_vectors = [soap_desc.create(step) for step in relevant_steps]
+
+        # Extract the SOAP vectors for all Li atom at each step
+        soap_vectors_all = [vectors[Li_indices] for vectors in soap_vectors]
+
+        # Convert the SOAP vectors to a numpy array
+        soap_vectors_array = np.array(soap_vectors_all)
+
+        # Save the SOAP vectors to a file
+        np.save(save_file, soap_vectors_array)
+
+        # Load the SOAP vectors from the .npy file
+        soap_vectors = np.load(save_file)
+
+        return soap_vectors
+
+    def _dim_reduction(self, vectors: list[np.ndarray], dim: int):
+        """Perform dimension reduction on the SOAP vectors."""
+
+        # Reshape the SOAP vectors into a 2D array
+        soap_vectors_2d = soap_vectors.reshape(soap_vectors.shape[0], -1)
+
+        # Perform PCA dimension reduction
+        pca = PCA(n_components=2)  # Set the desired number of components
+        reduced_vectors = pca.fit_transform(soap_vectors_2d)
+         
+        return reduced_vectors
+
+    def _cluster(self, data: list[Structure]):
+        """Perform DBSCAN."""
+        eps = self.db_kwargs.get('eps', 0.00019)
+        min_samples = self.db_kwargs.get('min_samples', 4)
+
+        db = DBSCAN(eps=eps, min_samples=min_samples).fit(data)
+        labels = db.labels_
+
+        return labels
+
     def sample(self, data: list[Structure]) -> list[Structure]:
```

Comment:
This is the overall function that you would finally call. This basically calls the above 6 steps. You need to finish this function. The logic is there, you just need to add a few stuff. 

In the end, you just need to call this function, and everything should be done. You need to test it. 

Line 255 is step 2 above. 

Line 270 needs to be filled.